### ISSUE TITLE
How to load a file with a record as a Map?

### ISSUE THREAD

**Issue by neongreen** [dhall-lang/dhall-haskell#1480]
I have a file with a record:

```
-- foo.dhall
{ a = "x", b = "y" }
```

What is the idiomatic way to load it as a `Map Text Text` in Haskell, without changing the `.dhall` file? `inputFile (Dhall.map strictText strictText) path` won't work, because `Dhall.map` doesn't work on records. Something like `input (Dhall.map strictText strictText) ("toMap ./" <> path)` will work, but constructing Dhall expressions like that is icky. I could construct a small piece of AST but it seems like overkill.

**Comment by Gabriel439** [comment]
@neongreen: Another option is to include `toMap` in `foo.dhall`

However, if that's not the option and the path is not statically known then I would go with "construct a small piece of AST"

**Comment by sjakobi** [comment]
If this is a common scenario, we could add utilities like

```
inputRecordAsMap :: Type v -> Text -> IO (Map Text v)
inputFileRecordAsMap :: Type v -> FilePath -> IO (Map Text v)
```

**Comment by Gabriel439** [comment]
@sjakobi: I think general utility to add would take an `Expr` as input:

```haskell
??? :: Type v -> Expr Src Import -> IO v
```

**Comment by neongreen** [comment]
What about `mapFromRecord :: Type v -> Type (Map Text v)`? Can this fly?

**Comment by sjakobi** [comment]
> What about `mapFromRecord :: Type v -> Type (Map Text v)`? Can this fly?
 
`Type` contains the `expected` type of the Dhall value, which in the case of records cannot be described just based on the type of the field value. To define the type of a record, you always need the fields of the record…

**Comment by neongreen** [comment]
> @sjakobi: I think general utility to add would take an `Expr` as input:
> 
> ```haskell
> ??? :: Type v -> Expr Src Import -> IO v
> ```

Ok, this would work perfectly.

**Comment by sjakobi** [comment]
> ```haskell
> ??? :: Type v -> Expr Src Import -> IO v
> ```

This should be

    ??? :: Type v -> Expr Src Import -> IO (Map Text v)

Right?

**Comment by Gabriel439** [comment]
@sjakobi: No, just `IO v`.  If you want to decode an `IO (Map Text v)` then you would provide it with a `Type (Map Text v)` (such as `auto`)

**Comment by neongreen** [comment]
Meanwhile I tried constructing "a small piece of AST", but got lost in parentheses:

```haskell
(Embed (Import (ImportHashed Nothing (Local Here (File ... I don't want to have to construct a 'File' too ...) Code)))
```

Luckily, I think [`getExpression`](https://hackage.haskell.org/package/dhall-1.27.0/docs/Dhall-Util.html#v:getExpression) will come to rescue:

```haskell
file <- getExpression (InputFile path)
??? (Dhall.map strictText strictText) (ToMap file)
```

**Comment by Gabriel439** [comment]
@neongreen: I still think the simplest solution is to wrap the record within the original file in `toMap`

**Comment by neongreen** [comment]
@Gabriel439 it sure is, but in my case I'm also using the file elsewhere (in other `.dhall` files) and I need to keep the record structure.

**Comment by neongreen** [comment]
I.e. here's what I have:

```dhall
-- people.dhall

{ alice = ...
, bob = ...
, josh = ...
}
```

```dhall
-- post1.dhall

let people = ./people.dhall

in Post::{
     , author = people.josh
     , ...
   }
```

I need to be able to use `people.dhall` as a record in `post1.dhall`, and simultaneously I need to be able to load `people.dhall` in Haskell.

**Comment by Gabriel439** [comment]
@neongreen: Could you create another file wrapping `./people.dhall`, like this:

```dhall
-- ./config.dhall
toMap ./people.dhall
```

**Comment by neongreen** [comment]
Certainly. (Feels like a workaround, though.)

**Comment by Gabriel439** [comment]
I just realized that the API already provides something fairly similar to the mystery `???` function which is `extract`:

```haskell
extract :: Type a -> Expr Src Void -> Extractor Src Void a
```

**Comment by neongreen** [comment]
`???` would also have to contain most of the [`inputWithSettings`](https://hackage.haskell.org/package/dhall-1.27.0/docs/src/Dhall.html#inputWithSettings) internals, where `extract` is only the last step.
